st2actions/st2actions/worker.py

```diff
@@ -134,7 +138,21 @@ def process(self, liveaction):
 
     def shutdown(self):
         super(ActionExecutionDispatcher, self).shutdown()
-        # Abandon running executions if incomplete
+
+        # Check if there are other runners in service registry
+        coordinator = coordination.get_coordinator()
+        member_ids = []
+        service = "actionrunner"
+
+        try:
+            member_ids = list(coordinator.get_members(service.encode("utf-8")).get())
+        except GroupNotCreated:
+            pass
+
+        while member_ids and self._running_liveactions:
```

Comment:
That sleep delay and code path in general is something which probably needs to be user configurable (and likely also have ability to fully disable that functionally via config option / feature flag).

And we may also want a hard timeout after which we give up - technically, depending on the service manager used (systemd, upstart) that may already be handled there (first send SIGTERM, if service still running after X seconds, send SIGKILL), but it's not a guarantee.

Reply:
Addressed